mithril-aggregator/src/artifact_builder/cardano_database_artifacts/digest.rs

```diff
@@ -0,0 +1,414 @@
+use std::{
+    fs,
+    path::{Path, PathBuf},
+    sync::Arc,
+};
+
+use anyhow::Context;
+use async_trait::async_trait;
+use mithril_common::{
+    entities::DigestLocation, logging::LoggerExtensions,
+    messages::CardanoDatabaseDigestListItemMessage, StdResult,
+};
+use reqwest::Url;
+use slog::{error, Logger};
+
+use crate::ImmutableFileDigestMapper;
+
+/// The [DigestFileUploader] trait allows identifying uploaders that return locations for digest archive files.
+#[cfg_attr(test, mockall::automock)]
+#[async_trait]
+pub trait DigestFileUploader: Send + Sync {
+    /// Uploads the archive at the given filepath and returns the location of the uploaded file.
+    async fn upload(&self, filepath: &Path) -> StdResult<DigestLocation>;
+}
+
+pub struct DigestArtifactBuilder {
+    /// Aggregator URL prefix
+    aggregator_url_prefix: Url,
+    /// Uploaders
+    uploaders: Vec<Arc<dyn DigestFileUploader>>,
+
+    digests_dir: PathBuf,
+
+    immutable_file_digest_mapper: Arc<dyn ImmutableFileDigestMapper>,
+
+    logger: Logger,
+}
+
+impl DigestArtifactBuilder {
+    /// Creates a new [DigestArtifactBuilder].
+    pub fn new(
+        aggregator_url_prefix: Url,
+        uploaders: Vec<Arc<dyn DigestFileUploader>>,
+        digests_dir: PathBuf,
+        immutable_file_digest_mapper: Arc<dyn ImmutableFileDigestMapper>,
+        logger: Logger,
+    ) -> StdResult<Self> {
+        Ok(Self {
+            aggregator_url_prefix,
+            uploaders,
+            digests_dir,
+            immutable_file_digest_mapper,
+            logger: logger.new_with_component_name::<Self>(),
+        })
+    }
+
+    pub async fn upload(&self) -> StdResult<Vec<DigestLocation>> {
+        let digest_path = self.create_digest_archive().await?;
+
+        let locations = self.upload_digest_archive(&digest_path).await;
+        fs::remove_file(&digest_path).with_context(|| {
+            format!(
+                "Could not remove digest archive file: '{}'",
+                digest_path.display()
+            )
+        })?;
+
+        locations
+    }
+
+    async fn create_digest_archive(&self) -> StdResult<PathBuf> {
```

Comment:
Is this the correct name, since we are creating a `.json` file? Perhaps something like:
```suggestion
    async fn create_digests_file(&self) -> StdResult<PathBuf> {
```